src/wiktextract/extractor/de/pronunciation.py

```diff
@@ -0,0 +1,201 @@
+from collections import defaultdict
+from typing import Dict, List, Union
+
+from wikitextprocessor import NodeKind, WikiNode
+from wikitextprocessor.parser import LevelNode
+from wiktextract.extractor.share import create_audio_url_dict
+
+from wiktextract.page import clean_node
+from wiktextract.wxr_context import WiktextractContext
+
+
+def extract_pronunciation(
+    wxr: WiktextractContext,
+    page_data: List[Dict],
+    level_node: LevelNode,
+):
+    for list_node in level_node.find_child(NodeKind.LIST):
+        sound_data = [defaultdict(list)]
+
+        for not_list_item_node in list_node.invert_find_child(
+            NodeKind.LIST_ITEM
+        ):
+            wxr.wtp.debug(
+                f"Found unexpected non-list-item node in pronunciation section: {not_list_item_node}",
+                sortid="extractor/de/pronunciation/extract_pronunciation/28",
+            )
+
+        for list_item_node in list_node.find_child(NodeKind.LIST_ITEM):
+            children = list_item_node.children
+            if len(children) == 0:
+                continue
+
+            head_template, rest = children[0], children[1:]
+            if (
+                not isinstance(head_template, WikiNode)
+                or head_template.kind != NodeKind.TEMPLATE
+                or not rest
+            ):
+                wxr.wtp.debug(
+                    f"Found unexpected non-template node in pronunciation section: {head_template}",
+                    sortid="extractor/de/pronunciation/extract_pronunciation/37",
+                )
+                continue
+            if head_template.template_name == "IPA":
+                process_ipa(wxr, sound_data, rest)
+            elif head_template.template_name == "Hörbeispiele":
+                sound_data.append(defaultdict(list))
+                process_hoerbeispiele(wxr, sound_data, rest)
+            elif head_template.template_name == "Reime":
+                process_rhymes(wxr, sound_data, rest)
+            else:
+                wxr.wtp.debug(
+                    f"Found unexpected template in pronunciation section: {head_template} with content {rest}",
+                    sortid="extractor/de/pronunciation/extract_pronunciation/45)",
+                )
+
+        # Remove empty entries
+        sound_data = [entry for entry in sound_data if entry != {}]
+        if len(sound_data) > 0:
+            page_data[-1]["sounds"].extend(sound_data)
+
+    for non_list_node in level_node.invert_find_child(NodeKind.LIST):
+        wxr.wtp.debug(
+            f"Found unexpected non-list node in pronunciation section: {non_list_node}",
+            sortid="extractor/de/pronunciation/extract_pronunciation/64",
+        )
+
+
+def process_ipa(
+    wxr: WiktextractContext,
+    sound_data: List[Dict],
+    nodes: List[Union[WikiNode, str]],
+):
+    if not nodes:
+        return
+
+    def process_lautschrift_template(
```

Comment:
Please don't use nested function unless you have to, local variables in these functions are often got overwritten accidentally. And again `TemplateNode.template_patameters` should be used here.

Reply:
Alright. I nested the function to indicate that this it is used only here and basically just for readability. 

I unnested it now.

Reply:
Sorry for forgetting to mention this is also a recommendation but not a rule. I just kind of scared by the long nested functions in the English extractor...